### PR TITLE
Code generation: first edition

### PR DESCRIPTION
**Pull request by xosmig**

### CODE REVIEW

File: codegen/README.md [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+- It forces the [“waterfall”](https://en.wikipedia.org/wiki/Waterfall_model) model on the development of modules. Since any change to the event and message descriptions (currently, in `.proto` files) needs to be propagated to all the related boilerplate, the programmer is discouraged from making such changes and, thus, is incentivized to think through the implementation and decide which events and network messages will be used before they start actually implementing the logic of the module. This approach is not always optimal for prototyping and exploring new ideas.
+
+The code generator aims to address these problems and make Mir a higher level framework. It may also simplify migration to a different model for representing events and network messages, or even allow multiple such models to coexist by performing an automatic conversion between them when necessary.
+
+## Usage
+
+### Annotating the `.proto` definitions
+
+See `protos/mir/codegen_extensions.proto`, `protos/dsl/codegen_extensions.proto`, and `protos/net/codegen_extensions.proto` for the list of extensions.
```
Comment (matejpavlovic) [inline codegen/README.md:17]: `protos/dsl/codegen_extensions.proto` seems to not be there. (Also mentioned at other places in this file.)

File: codegen/README.md [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+The code generator should be typically run using [go generate](https://go.dev/blog/generate), [GNU make](https://www.gnu.org/software/make/), a bash script, or another similar tool.
+The following order is important to avoid circular dependencies (the examples are in the [go generate](https://go.dev/blog/generate) format):
+
+1. Compile the protbuf extensions:
+    ```go
+    //go:generate -command protoc-basic protoc --proto_path=. --go_out=../pkg/pb/ --go_opt=paths=source_relative
+
+    // Generate the code for codegen extensions.
+    //go:generate protoc-basic mir/codegen_extensions.proto
+    //go:generate protoc-basic dsl/codegen_extensions.proto
```
Comment (matejpavlovic) [inline codegen/README.md:166]: dsl/codegen_extensions.proto

File: codegen/README.md [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+For a set of reasons (including flexibility, ease of implementation, and potential future migration to a different representation) the Mir code generator is *not* implemented as a protoc plugin.
+Instead, it uses the Go code generated by `protoc` as its input.
+It inspects this code using reflection.
+Hence, the whole package containing the `protoc`-generated code must be compilable (i.e., contain no syntax errors).
+
+We proceed by describing each of the components one by one.
+
+### The protobuf extensions
+
+Located in `protos/mir/codegen_extensions.proto`, `protos/dsl/codegen_extensions.proto`, and `protos/net/codegen_extensions.proto`.
```
Comment (matejpavlovic) [inline codegen/README.md:211]: dsl/codegen_extensions.proto

**Review by matejpavlovic — APPROVED** [review]
Great work Andrei!

I did not go through the code line by line, but I used it to implement the ping-pong protocol from scratch and it works seamlessly.

I only left a few comments on the readme (about the same thing) that should be a no-brainer to fix.

Thanks!

File: codegen/README.md (hunk 1 above)
Comment (xosmig) [reply codegen/README.md]: Ah, thank you. This file used to exist but was removed. Fixed

File: codegen/README.md (hunk 2 above)
Comment (xosmig) [reply codegen/README.md]: fixed

File: codegen/README.md (hunk 3 above)
Comment (xosmig) [reply codegen/README.md]: fixed

**Pull request merged**